src/ruby/puppetserver-lib/puppet/server/master.rb

```diff
@@ -69,6 +75,8 @@ def handleRequest(request)
         body_to_return,
         response[:content_type],
         response["X-Puppet-Version"])
+  ensure
+    Puppet.settings.clear_local_settings if Puppet.settings.is_a?(Puppet::Server::Settings)
```

Comment:
One thing that might make sense here is to change this to check for a `multithreaded` instance variable that we could set during initialization. I don't know think it's much of a real difference, but I could change it if people had a preference.

Reply:
hmm, checking for `multithreaded` seems more "right", but not sure if it's worth an extra variable? maybe you could just add a comment here that this is done for thread safety?